### ISSUE TITLE
Failed adv training: MNIST and 2_layer_MLP

### ISSUE THREAD

**Issue by TianQi-777** [BorealisAI/advertorch#54]
Hello, I used your code to train 2 layers of MLP and found that I couldn't adv train successfully.  My modified code is as follows:
```
class SimpleModel(nn.Module):
    def __init__(self, dim_input=DIM_INPUT, num_classes=NUM_CLASS):
        super(SimpleModel, self).__init__()
        self.fc1 = nn.Linear(dim_input, 10)
        self.fc2 = nn.Linear(10, num_classes)

    def forward(self, x):
        x = x.view(x.size(0), -1) # MNIST input
        x = self.fc1(x)
        x = F.relu(x)
        x = self.fc2(x)
        return x
```  
The result is that `cln acc` and `adv acc` are always 10%
I find that the gradient disappears  and the predictions of the adversarial samples are always the same class, do you know why?(I tried adjusting the learning rate but it didn't seem to work, and the code is correct during standard training)
I also tested the `MLP` in your implementation, but still cannot achieve adversarial training(MNIST)
Looking forward to your reply

**Comment by gwding** [comment]
Hi @TianQi-777 , are you using this file https://github.com/BorealisAI/advertorch/blob/master/advertorch_examples/tutorial_train_mnist.py ?

the 10% acc, is it only on test? or also for training?

Adversarial training usually requires a large model to work, 10 hidden units definitely looks way too small to me. I'm not sure whether clean training would work on such a small network.

If training a small network is your goal, I would suggest gradually reduce the size of the model and see how it works?

**Comment by TianQi-777** [comment]
Thank you for your reply.
I did use the file`tutorial_train_mnist.py`.
Well, we do not consider such a small network first, I also tested the tutorial network `MLP` you gave, it is located in [advertorch/test_utils.py](https://github.com/BorealisAI/advertorch/blob/5daeeee98077afacf94728936a04cebde474259f/advertorch/test_utils.py#L108). From its input `self.linear1 = nn.Linear(28 * 28, 300)`, it looks like it was prepared for MNIST. I modified two lines of [tutorial_train_mnist.py](https://github.com/BorealisAI/advertorch/blob/5daeeee98077afacf94728936a04cebde474259f/advertorch_examples/tutorial_train_mnist.py#L53) code, as shown below:
```
from advertorch.test_utils import MLP
model = MLP()                                                         # Change Line 53 (Orignal: model = LeNet5())
```
Unfortunately, the result is
```
Train Epoch: 0 [0/60000 (0%)]	Loss: 2.546685
Train Epoch: 0 [10000/60000 (17%)]	Loss: 2.306097
Train Epoch: 0 [20000/60000 (33%)]	Loss: 2.299403
Train Epoch: 0 [30000/60000 (50%)]	Loss: 2.308362
Train Epoch: 0 [40000/60000 (67%)]	Loss: 2.302292
Train Epoch: 0 [50000/60000 (83%)]	Loss: 2.302776

Test set: avg cln loss: 2.3029, cln acc: 1010/10000 (10%)

Test set: avg adv loss: 2.3034, adv acc: 1009/10000 (10%) best_adv acc: 0.0000

Train Epoch: 1 [0/60000 (0%)]	Loss: 2.295106
Train Epoch: 1 [10000/60000 (17%)]	Loss: 2.302987
Train Epoch: 1 [20000/60000 (33%)]	Loss: 2.295572
Train Epoch: 1 [30000/60000 (50%)]	Loss: 2.293433
Train Epoch: 1 [40000/60000 (67%)]	Loss: 2.308897
Train Epoch: 1 [50000/60000 (83%)]	Loss: 2.303285

Test set: avg cln loss: 2.3020, cln acc: 1010/10000 (10%)

Test set: avg adv loss: 2.3024, adv acc: 1010/10000 (10%) best_adv acc: 10.0900

Train Epoch: 2 [0/60000 (0%)]	Loss: 2.306617
Train Epoch: 2 [10000/60000 (17%)]	Loss: 2.303803
Train Epoch: 2 [20000/60000 (33%)]	Loss: 2.297284
Train Epoch: 2 [30000/60000 (50%)]	Loss: 2.297552
Train Epoch: 2 [40000/60000 (67%)]	Loss: 2.302041
Train Epoch: 2 [50000/60000 (83%)]	Loss: 2.294422

Test set: avg cln loss: 2.3014, cln acc: 1010/10000 (10%)

Test set: avg adv loss: 2.3018, adv acc: 1010/10000 (10%) best_adv acc: 10.1000
```
This looks like training failed. I think it may be that the MLP network is not easy to conduct adversarial training? After all there are convolution units in [LeNet](https://github.com/BorealisAI/advertorch/blob/5daeeee98077afacf94728936a04cebde474259f/advertorch/test_utils.py#L85)

**Comment by gwding** [comment]
The `MLP` class in `advertorch/test_utils.py` is actually only for testing purposes. I guess a larger MLP with batch normalization might work better, but personally never tried.

**Comment by TianQi-777** [comment]
OK, thank you for your reply, this should be the problem of adversarial training itself. I tried to implement this idea in other frameworks, but it still doesn't work, so it should be a problem with the algorithm itself. Anyway, `Advertorch` is a great project.
